keras/src/layers/preprocessing/image_preprocessing/equalization.py

```diff
@@ -0,0 +1,193 @@
+
+    Histogram equalization is a technique to adjust image intensities to
+    enhance contrast by effectively spreading out the most frequent
+    intensity values. This layer applies equalization on a channel-wise
+    basis, which can improve the visibility of details in images.
+
+    This layer works with both grayscale and color images, performing
+    equalization independently on each color channel. At inference time,
+    the equalization is consistently applied.
+
+    **Note:** This layer is safe to use inside a `tf.data` pipeline
+    (independently of which backend you're using).
+
+    Args:
+        value_range: Optional list/tuple of 2 floats specifying the lower
+            and upper limits of the input data values. Defaults to `[0, 255]`.
+            If the input image has been scaled, use the appropriate range
+            (e.g., `[0.0, 1.0]`). The equalization will be scaled to this
+            range, and output values will be clipped accordingly.
+        bins: Integer specifying the number of histogram bins to use for
+            equalization. Defaults to 256, which is suitable for 8-bit images.
+            Larger values can provide more granular intensity redistribution.
+
+    Inputs: 3D (HWC) or 4D (NHWC) tensor, with float or int dtype. Input pixel
+        values can be of any range (e.g. `[0., 1.)` or `[0, 255]`)
+
+    Output: 3D (HWC) or 4D (NHWC) tensor with intensities redistributed to
+        enhance contrast. The output will be of the same shape and dtype as
+        the input, with values scaled and clipped to the specified
+        `value_range`.
+
+    Example:
+
+    ```python
+    # Create an equalization layer for standard 8-bit images
+    equalizer = keras.layers.Equalization()
+
+    # An image with uneven intensity distribution
+    image = [...] # your input image
+
+    # Apply histogram equalization
+    equalized_image = equalizer(image)
+
+    # For images with custom value range
+    custom_equalizer = keras.layers.Equalization(
+        value_range=[0.0, 1.0],  # for normalized images
+        bins=128  # fewer bins for more subtle equalization
+    )
+    custom_equalized = custom_equalizer(normalized_image)
+    ```
+    """
+
+    def __init__(self, value_range=(0, 255), bins=256, **kwargs):
+        super().__init__(**kwargs)
+        self.bins = bins
+        self._set_value_range(value_range)
+
+    def _set_value_range(self, value_range):
+        if not isinstance(value_range, (tuple, list)):
+            raise ValueError(
+                self._VALUE_RANGE_VALIDATION_ERROR
+                + f"Received: value_range={value_range}"
+            )
+        if len(value_range) != 2:
+            raise ValueError(
+                self._VALUE_RANGE_VALIDATION_ERROR
+                + f"Received: value_range={value_range}"
+            )
+        self.value_range = sorted(value_range)
+
+    def _custom_histogram_fixed_width(self, values, value_range, nbins):
```

Comment:
Can `keras.ops.histogram` be used?

Reply:
1. using `self.backend.numpy.histogram` raises : AttributeError: 'SymbolicTensor' object has no attribute 'numpy' because of this line 
https://github.com/keras-team/keras/blob/dbf9878e2c64b92e15b8a221cf999d7b7901c03d/keras/src/backend/tensorflow/numpy.py#L2627

2. Also `self.backend.numpy.histogram` doesnt cover some of the edge cases like it can handle extreme values using clipping method, handles values outside the specified range
3. also i have added scaling approach to map values to bin indices